---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.


############ script ################
# libraries
```{r}
library(readxl)
library(dplyr)
library(GenomicRanges)
library(Gviz)
library(shiny)
```

# defaults
```{r}
from <- 95433599
to <- 95579959  #95579759
genome <- "hg38"
chr <- "chr13"
```

# 1  GVIZ plot of the whole Iso A + B
## 1a -first 2 tracks (Ideogram, GenomicAxis)
```{r}
itrack <- IdeogramTrack(genome = genome, chromosome = chr)
gtrack <- GenomeAxisTrack()

plotTracks(list(itrack,gtrack), showBandId = TRUE, cex.bands = 0.7,from = from, to = to)
```



## 1b -CpGs track
```{r}
# a- get the final table for all cpgs
cpgs <- read_excel("./Output/10-cpgs_for_Granges.xlsx",col_names = TRUE)
View(cpgs)

# b- convert to genomicRanges
gr_cpgs <- GRanges(seqnames = cpgs$chr,
              ranges = IRanges(start = cpgs$pos, end = cpgs$pos)) #,strand = cpgs_AB$strand)
mcols(gr_cpgs) <- cpgs[, 4:ncol(cpgs)]
gr_cpgs


##### please replace the GenType from the beginneing in the dmp pipeline
# Define the colors for GenType
#fill_colors <- c("Other"= "lightblue" ,"CLDN10_IsoA" = "blue", "Targeted_CpG" = "red", "CLDN10_IsoB" = "green")

# Create AnnotationTrack with CpGs
sondentrack <- AnnotationTrack(gr_cpgs, chromosome = chr, genome = genome , name = "CpG Probes", 
                          col = NA,       # Border color
                          #fill = fill_colors[gr_cpgs_AB$GenType], 
                          id = gr_cpgs$CpGs,
                          feature = gr_cpgs$GenType)  

# Create the plot
#plotTracks(list(itrack, gtrack, sondentrack), from = from, to = to, showBandId = TRUE,cex.bands = 0.7, "Other"= "lightblue", "CLDN10_IsoA" = "blue", "Targeted_CpG" = "red", "CLDN10_IsoB" = "green",stacking = "dense")
```

## 1c -beta values pointplot track
```{r}
# the strand has to be unique here
#data = must be a data matrix
#from the previus step we have gr_cpgs

# Extract beta values from metadata columns (as DataFrame)
beta_values <- as.data.frame(mcols(gr_cpgs)[, tail(seq_along(mcols(gr_cpgs)), 8)])


betaTrack <- DataTrack(gr_cpgs, chromosome = chr, genome = genome,
                    name = "Beta values",
                    data = beta_values)
 
#plotTracks(betaTrack, groups = rep(c("unguided", "guided"), each = 4),
#           type = c("a", "p"), from = from , to = to) # or with aggregate aggregateGroups = TRUE, aggregation = "mean"
```

## 1  -Offtarget track
```{r}
# the granges of the offtarget is from script 11 
#gr_offtargets

#pos on chr13 wo offtarget gibt
von = 113103059
zu =113104000

offtargetTrack <- AnnotationTrack(gr_offtargets, genome = genome, chromosome = chr,
                                  name = "potential Off-targets\n[-70,+70]", feature = paste(gr_offtargets$mcols.guide, "\n", gr_offtargets$mcols.Number_of_mismatches, "mismatches"), fill = "red")

#plotTracks(offtargetTrack, chromosome = "chr13", from = von, to= zu, featureAnnotation ="feature", fontcolor.feature= "black")
```

## 1E -DMR track
### create DMR Granges 
```{r}
##### change the DMRnR FROM THE BEGINNING IN THE THE PIPELINE
DMRs <- read_excel("./Output/8-DMRs_0.01_smooth.xlsx", col_names = TRUE)

DMRs$DMR_Nr <- paste("DMR", DMRs$DMR_Nr, sep = "")

View(DMRs)

# check if all numbers are numeric
DMRs <- DMRs %>%
  mutate(across(c(start, end, length, p.value), as.numeric))

# create a granges of the DMRs 
gr_Dmrs <- GRanges(
  seqnames = DMRs$chr,  
  ranges = IRanges(start = DMRs$start, end = DMRs$end),
  feature = DMRs$DMR_Nr,
  p.value = DMRs$p.value
)
#gr_Dmrs


DmrTrack <- AnnotationTrack(gr_Dmrs,genome = genome ,chromosome = chr, name = "DMRs", col = "pink", fill = "pink",feature = paste(gr_Dmrs$feature, " P.val=", round(gr_Dmrs$p.value, 5))) # delete pvalue if he dmr table is filtered

#plotTracks(DmrTrack,from = from, to = to, groupAnnotation= "feature")
```


## 1F -TAD Track CAKI2
```{r}
# create TAD Granges
tissue <- "CAKI2"
binsize <- 25 # or 10


TADs <- read_excel(sprintf("../PythonProject/TADs_CAKI2/3_TADs_as_txt_and_Excel/%s_%s_%skb_TADs.xlsx",tissue, chr, binsize), skip =2)

# Convert columns to numeric
TADs <- TADs %>%
  mutate(across(c(StartPos, EndPos, TadLength), as.numeric))


# convert Tad_df to Granges 
gr_TADs <- GRanges(seqnames = chr,
              ranges = IRanges(start = TADs$StartPos, end = TADs$EndPos))
mcols(gr_TADs) <- TADs[, 1]
#gr_TADs

### plot tad track
TadTrack <- AnnotationTrack(gr_TADs , genome = genome, 
                            name = paste0("TADs\nCaki2\n",binsize, "kb"), 
                            feature = gr_TADs$TadNr,  
                            col = "black")

plotTracks(TadTrack, groupAnnotation = "feature", from = from, to = to) # try to change shape 
```

## 1g -SUBTADs Track CAKI2
```{r}
# create TAD Granges
tissue <- "CAKI2"
chr <- "chr13"
binsize <- 25
from <- 95433599
to <- 95579759

SUBTADs_filepath <- sprintf("../PythonProject/TADs_CAKI2/3_TADs_as_txt_and_Excel/%s_%s_%skb_SubTADs_noDuplicates.txt", tissue, chr, binsize)
SUBTADs <- read.delim(SUBTADs_filepath, header = TRUE)
View(SUBTADs) #80


# convert to Granges 
gr_SUBTADs <- GRanges(seqnames = chr,
              ranges = IRanges(start = SUBTADs$StartPos, end = SUBTADs$EndPos))
gr_SUBTADs

### plot tad track
SUBTadTrack <- AnnotationTrack(gr_SUBTADs, genome = genome, 
                            name = paste0("SUBTADs\nCaki2\n",binsize, "kb"), col= "black")

#plotTracks(SUBTadTrack, from = from, to = to) # try to change shape 
```


## 1g -TAD Track SKIN -> no files of skin in this dir
### 10kb create TAD Granges
```{r}
tissue2 <- "skin"
chr <- "chr13"
binsize <- 25
from <- 95433599
to <- 95579759

#SUBTADs_filepath2 <- sprintf("../PythonProject/TADs_CAKI2/3_TADs_as_txt_and_Excel/%s_%s_%skb_SubTADs_noDuplicates.txt", tissue2, chr, binsize)
#SUBTADs2 <- read.delim(SUBTADs_filepath, header = TRUE)
#View(SUBTADs2)
# Convert columns to numeric
tad_skin_10kb <- tad_skin_10kb %>%
  mutate(across(c(StartPos, EndPos, TadLength), as.numeric))


# convert Tad_df to Granges 
chr_TAD <- "chr13"
gr_tad_skin_10kb <- GRanges(seqnames = chr_TAD,
              ranges = IRanges(start = tad_skin_10kb$StartPos, end = tad_skin_10kb$EndPos))
mcols(gr_tad_skin_10kb) <- tad_skin_10kb[, 1]
gr_tad_skin_10kb


### plot tad track
TadTrack_skin_10kb <- AnnotationTrack(gr_tad_skin_10kb, 
                            name = paste0("TADs-Skin fibroblast-",bin_size_skin10), 
                            feature = gr_tad_skin_10kb$TadNr,  
                            col = "black")

plotTracks(TadTrack_skin_10kb,groupAnnotation = "feature", from = from, to = to)
```

### 25kb create TAD Granges
```{r}
bin_size_skin25 <- 25000
tad_skin_25kb <- read_excel("C:/Users/ghaza/Documents/ghazal/Bioinformatik_FÃ¤cher/Masterarbeit_Project/Scripts/PythonProject/TADs_skin/3_TADs_as_txt_and_Excel/Skin_fibro_chr13_25000bp_TADs.xlsx", skip =2)

# Convert columns to numeric
tad_skin_25kb <- tad_skin_25kb %>%
  mutate(across(c(StartPos, EndPos, TadLength), as.numeric))


# convert Tad_df to Granges 
chr_TAD <- "chr13"
gr_tad_skin_25kb  <- GRanges(seqnames = chr_TAD,
              ranges = IRanges(start = tad_skin_25kb$StartPos, end = tad_skin_25kb$EndPos))
mcols(gr_tad_skin_25kb) <- tad_skin_25kb[, 1]
gr_tad_skin_25kb 


### plot tad track
TadTrack_skin_25kb <- AnnotationTrack(gr_tad_skin_25kb , 
                            name = paste0("TADs-Skin fibroblast-",bin_size_25), 
                            feature = gr_tad_skin_25kb$TadNr,  
                            col = "black")

plotTracks(TadTrack_skin_25kb,groupAnnotation = "feature", from = from, to = to)
```



## 1G -CPG Islands track
```{r}
# GRanges of CpGIslands

#library(AnnotationHub) #dont work-> all on it are hg19

# download cpgislands from ucsc
url <- "http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cpgIslandExt.txt.gz"
destfile <- "cpgIslandExt_hg38.txt.gz"

if (!file.exists(destfile)) {
  download.file(url, destfile)
  message("File downloaded.")
} else {
  message("cpgIslandExt_hg38.txt.gz File already exists. Skipping download.")
}
 #701 KB

# Read the file (it's gzipped, so use gzfile)
cpg_Islands <- read.table(gzfile(destfile), header = FALSE, sep = "\t", stringsAsFactors = FALSE)

colnames(cpg_Islands) <- c("bin", "chrom","chromStart","chromEnd","name","length","cpgNum","gcNum","perCpg","perGc", "obsExp")
#View(cpg_Islands)

#Note: all start coordinates in our database are 0-based, not 1-based. 

# convert to GRanges
gr_cpgIslands <- GRanges(
  seqnames = cpg_Islands$chrom,   
  ranges = IRanges(
    start = cpg_Islands$chromStart + 1,  # Convert from 0-based to 1-based
    end = cpg_Islands$chromEnd))

mcols(gr_cpgIslands) <- cbind(
  feature = cpg_Islands$name,
  cpg_Islands[, 7:ncol(cpg_Islands)]
)
#gr_cpgIslands

# make AnnotationTrack
cpgIslandTrack <- AnnotationTrack(gr_cpgIslands,
  chromosome = chr,
  genome = genome,
  name = "CpG Islands",
  feature = mcols(gr_cpgIslands)$feature,
  fill = "darkgreen")

#plotTracks(cpgIslandTrack, from = from, to = to, groupAnnotation = "feature",just.group = "right")
```


## 1H -Gene track
### 1H.1 from UCSC -> no gene names
```{r}
# you can get a TxDb object from any GFF  file easily 
#BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(TxDb.Hsapiens.UCSC.hg38.knownGene) #--> anstatt knownGene -> refGene

TxDb <-  TxDb.Hsapiens.UCSC.hg38.knownGene
#TxDb
#unique(feature(GeneTrack))

# ploting
ensTransTrack <- GeneRegionTrack(TxDb, chromosome = chr, name = "transcript", start = from, end = to, transcriptAnnotation = "transcript", just.group = "above")

#plotTracks(ensTransTrack)
```


### 1H.2 from ENsDb: cant pack gene, transcript and exon infos in one granges and plot gene cause it is will be replicated according to the numner of transcripts/ exons bzw.rows
```{r}
#BiocManager::install("EnsDb.Hsapiens.v86")  # Ensembl 86 (GRCh38)

library(EnsDb.Hsapiens.v86)
library(Gviz)
library(GenomicRanges)
library(GenomeInfoDb)

# Get EnsDb
edb <- EnsDb.Hsapiens.v86

#To extract specific data, such as transcript or gene information, from an EnsDb object, you can use functions like transcripts(), genes(), and exons().

tx_gr <- transcripts(edb, columns = c("gene_id", "gene_name", "tx_id", "tx_biotype", "exon_id"))
tx_gr
length(tx_gr) #1297863
seqlevelsStyle(tx_gr) <- "UCSC"

# Filter to standard chromosomes only????
tx_gr_filter <- keepSeqlevels(tx_gr, standardChromosomes(tx_gr), pruning.mode = "coarse")
length(tx_gr_filter) #1194162 --> 92.00987%

genEnsTrack <- AnnotationTrack(tx_gr_filter,
                          chromosome = chr,
                          name = "Gene Ens",
                          id= tx_gr_filter$gene_name,
                          feature = tx_gr_filter$tx_id,
                          group = tx_gr_filter$exon_id)
plotTracks(genEnsTrack, from = from, to = to, groupAnnotation= "id", just.group= "above")
```


### 1H.3 Get gene ranges (as GRanges) 
```{r}
#BiocManager::install("EnsDb.Hsapiens.v86")  # Ensembl 86 (GRCh38)
library(EnsDb.Hsapiens.v86)
library(Gviz)
library(GenomicRanges)
library(GenomeInfoDb)

# Get EnsDb
edb <- EnsDb.Hsapiens.v86

#To extract specific data, such as transcript or gene information, from an EnsDb object, you can use functions like transcripts(), genes(), and exons().
######################### Get gene ranges
tx_gr <- genes(edb)
#length(tx_gr) # 63970

# Convert to UCSC style 13 --> chr13
seqlevelsStyle(tx_gr) <- "UCSC"
#head(seqlevels(tx_gr))
#table(seqnames(tx_gr))["chr13"] 

# Filter to standard chromosomes only????
tx_gr_filtered <- keepSeqlevels(tx_gr, standardChromosomes(tx_gr), pruning.mode = "coarse")
#length(tx_gr_filtered)#58650

# Subset to just this region
#tx_gr_sub <- subset(tx_gr, seqnames == chr & start(tx_gr) < to & end(tx_gr) > from)
#head(tx_gr_sub)

# Make the GeneRegionTrack manually
geneTrack <- AnnotationTrack(tx_gr_filtered,
                          chromosome = chr,
                          name = "Gene Ens",
                          id= tx_gr_filtered$gene_name)

plotTracks(geneTrack, from = from, to = to, featureAnnotation= "id", fontcolor.feature = "darkblue")
```



### 1H.4 Get transcript ranges (as GRanges) --> not nice to see
```{r}
transcript_tx_gr <- transcripts(edb)
length(transcript_tx_gr) #  216741

# Convert to UCSC style 13 --> chr13
seqlevelsStyle(transcript_tx_gr) <- "UCSC"
head(seqlevels(transcript_tx_gr))
table(seqnames(transcript_tx_gr))["chr13"] 


# Filter to standard chromosomes only
transcript_tx_gr_filt <- keepSeqlevels(transcript_tx_gr, standardChromosomes(transcript_tx_gr), pruning.mode = "coarse")
# Check again
length(transcript_tx_gr_filt)# 198739 -->91.69423%

transcript_tx_gr_filt


# Subset to just this region
#tx_gr_sub <- subset(tx_gr, seqnames == chr & start(tx_gr) < to & end(tx_gr) > from)
#head(tx_gr_sub)

# Make the GeneRegionTrack manually
transTrack <- AnnotationTrack(transcript_tx_gr_filt,
                          chromosome = chr,
                          name = "Gene Ens",
                          id= transcript_tx_gr_filt$tx_id)

plotTracks(transTrack, from = from, to = to, groupAnnotation= "id", just.group= "above")
```




## 1 - Plot all tracks together + save
```{r}
# Suppress labels on atrack
#displayPars(atrack) <- list(groupAnnotation = "none",showFeatureId = FALSE)
# Set groupAnnotation = "feature" ONLY on the track(s) where you want to see feature labels

plotGvizTracks <- function(from, to){
    displayPars(DmrTrack) <- list(groupAnnotation = "feature")  
    displayPars(cpgIslandTrack) <- list(groupAnnotation = "feature") 
    displayPars(sondentrack) <- list(stacking= "dense") 
    displayPars(geneTrack) <- list(featureAnnotation= "id",fontcolor.feature = "darkblue") 
    displayPars(offtargetTrack) <- list(featureAnnotation ="feature", fontcolor.feature= "black") 

    plotTracks(list(itrack, gtrack, sondentrack, DmrTrack, cpgIslandTrack, offtargetTrack,dTrack,geneTrack,ensTransTrack,SUBTadTrack,TadTrack),
    from = from, to = to,
    showBandId = TRUE, cex.bands = 0.7,
    groups = rep(c("unguided", "guided"), each = 4),type = c("a", "p"),
    transcriptAnnotation = "transcript",
    cex.group = 0.9,
    cex.legend = 0.9, 
    col.title = "black",
    "Other" = "lightblue",
    "CLDN10_IsoA" = "blue",
    "Targeted_CpG" = "red",
    "CLDN10_IsoB" = "green"
    )
}


# save as a PDF file
#pdf("./Plots/GVIZ_CLDN10_AB_14_subtad25kb.pdf", width = 27, height = 15)  # dimensions in inches
#plotGvizTracks(95433599, 95579759)
#dev.off() 
```


# shiny app
```{r}
# Example ranges for DMRs (replace with your actual ones)
dmr_ranges <- list(
  DMR1 = c(95433599, 95579759),
  DMR2 = c(von, zu),
  DMR3 = c(96000000, 96150000)
)

chr13_length <- 114364328

ui <- fluidPage(
  titlePanel("Interactive Gviz Plot"),
  sidebarLayout(
    sidebarPanel( width = 2,
      selectInput("dmr", "Choose DMR:", choices = names(dmr_ranges), selected = "DMR1"),
      #sliderInput("range", "Genomic Range on chr13:",min = 1, max = chr13_length,value = dmr_ranges[["DMR1"]],step = 25000)
      br(), br(),
      numericInput("from", "From:", value = dmr_ranges[["DMR1"]][1], min = 1, max = chr13_length),
      numericInput("to", "To:", value = dmr_ranges[["DMR1"]][2], min = 1, max = chr13_length),
      actionButton("update", "Submit"),
      br(), br(), br(),
      downloadButton("downloadPlot", "Save Plot as PDF")
    ),
    mainPanel( width= 10,
      plotOutput("gvizPlot", height = "800px")
    )
  )
)

server <- function(input, output, session) {
  selectedRange <- reactiveVal(dmr_ranges[["DMR1"]])

  observeEvent(input$update, {
    selectedRange(c(input$from, input$to))
})

  
  # Update range input when DMR changes
  observeEvent(input$dmr, {
    updateNumericInput(session, "from", value = dmr_ranges[[input$dmr]][1])
    updateNumericInput(session, "to", value = dmr_ranges[[input$dmr]][2])
  })

  # Main plot output
  output$gvizPlot <- renderPlot({
    req(selectedRange())
    plotGvizTracks(selectedRange()[1], selectedRange()[2])
  })

  # Download handler for PDF
  output$downloadPlot <- downloadHandler(
    filename = function() {
      paste0("GvizPlot_chr13_", input$from, "_", input$to, ".pdf")
    },
    content = function(file) {
      pdf(file, width = 25, height = 15)
      plotGvizTracks(as.numeric(input$from), as.numeric(input$to))
      dev.off()
    }
  )
}

shinyApp(ui, server)
```