---
title: "the final plot"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.


######################## SCRIPT ####################




# retreave tables
```{r}
library(readxl)

dmrs_smooth <- read_excel("./Output/DMRs_0.01_smooth.xlsx")
Annotation_df <-read_excel("./Output/4-annotation_0.01_with_betas.xlsx", col_names = TRUE) #891492     51 
dmrsInTads <- read_excel("./Output/9-DMRs_inTADs.xlsx", col_names = TRUE)
```


# CLDN10 CpGs sortieren
```{r}
library(ggplot2)
library(plotly)
library(htmlwidgets)
library(dplyr)


### cpgs in the range of both variants A and B
# filter all cpgs both ranges of A and B
cpgs_inCLDN10_AB <- annotation_df %>%   filter(chr == "chr13" & pos >= 95433599 & pos <= 95579759)

# View the filtered table
View(cpgs_inCLDN10_AB) # 71 cpgs   42 columns


# Delete unnecessary columns
cpgs_inCLDN10_AB <- cpgs_inCLDN10_AB %>% select(-c(5, 6, 9:17, 20:24, 26,27, 34, 35))
View(cpgs_inCLDN10_AB)


### CLDN10 -B 
CLDN10_B_CPGs <- c("cg24426691_TC11", "cg16232183_TC21", "cg06428163_BC21", "cg16275739_TC11", "cg04603730_TC11", "cg10305311_BC11","cg09469554_BC11", "cg25032595_BC11", "cg16556145_BC11", "cg18393747_BC21" )


# Add column GenType
cpgs_inCLDN10_AB$GenType <- ifelse(
  cpgs_inCLDN10_AB$Name %in% CLDN10_B_CPGs, "Targeted CpG",
  ifelse(cpgs_inCLDN10_AB$pos > 95552720, "CLDN10 Iso-B", "CLDN10 Iso-A")
)

# Ensure GenType is a factor
cpgs_inCLDN10_AB$GenType <- as.factor(cpgs_inCLDN10_AB$GenType)

View(cpgs_inCLDN10_AB) # 71  23 -> changed name to cpgs_inCLDN10_AB
```


# Add the DMR infos to the targeted region cpgs
```{r}

# Step 1: Filter DMRs to only those on chromosome 13
dmrs_chr13 <- dmrsInTads[dmrsInTads$chr == "chr13", ]

# Step 2: Create new columns in the CpG table
cpgs_inCLDN10_AB <- cpgs_inCLDN10_AB %>%
  mutate(
    DmrNr = NA, DmrStart = NA, DmrEnd = NA, Dmrpval = NA, 
    )

# Step 3: Loop through CpGs and check if they fall into a DMR
for (i in 1:nrow(cpgs_inCLDN10_AB)) {
  pos <- cpgs_inCLDN10_AB$pos[i]
  
  # Find matching DMR(s)
  match <- dmrs_chr13[dmrs_chr13$start <= pos & dmrs_chr13$end >= pos, ]

  # If a match is found, assign values (only first match to avoid size conflict)
  if (nrow(match) > 0) {
    cpgs_inCLDN10_AB$DmrNr[i] <- match$DMR_Nr[1]
    cpgs_inCLDN10_AB$DmrStart[i] <- match$start[1]
    cpgs_inCLDN10_AB$DmrEnd[i] <- match$end[1]
    cpgs_inCLDN10_AB$Dmrpval[i] <- match$p.value[1]
  }
}

View(cpgs_inCLDN10_AB)
```


# Add TAD infos
```{r}
#tad_df <- read_excel("./Output/TAD_results/TADs_with_header.xlsx", skip = 2 ,col_names = TRUE)

cpgs_inCLDN10_AB <- cpgs_inCLDN10_AB %>%
  mutate(
   TadNr = NA, TadStart = NA, TadEnd = NA, DmrOverlap_withTad = NA
  )

for (i in 1:nrow(cpgs_inCLDN10_AB)) {
  pos <- cpgs_inCLDN10_AB$pos[i]
  match <- tad_df[tad_df$StartPos <= pos & tad_df$EndPos >= pos, ]
  
  if (nrow(match) > 0) {
    cpgs_inCLDN10_AB$TadNr[i] <- match$TadNr[1]
    cpgs_inCLDN10_AB$TadStart[i] <- match$StartPos[1]
    cpgs_inCLDN10_AB$TadEnd[i] <- match$EndPos[1]
  }
}

View(cpgs_inCLDN10_AB)     
```

# Save/retreave Excel table 6-cpgs_inCLDN10_A+B.xlsx
```{r}
cpgs_inCLDN10_AB_with_rownames <- cbind(CpGs = rownames(cpgs_inCLDN10_AB), cpgs_inCLDN10_AB)
writexl::write_xlsx(cpgs_inCLDN10_AB_with_rownames, "./Output/6-cpgs_inCLDN10_A+B.xlsx")

library(readxl)
cpgs_inCLDN10_AB <- read_excel("./Output/6-cpgs_inCLDN10_A+B.xlsx")
View(cpgs_inCLDN10_AB)

#  convert numbers to numeric
cpgs_inCLDN10_AB <- cpgs_inCLDN10_AB %>%
  mutate(
    Dmrpval = as.numeric(Dmrpval),
    DMR_label = ifelse(
      is.na(Dmrpval), 
      NA,  
      paste0("DMR", DmrNr, "-", round(Dmrpval, 5))
    ),
    pos = as.numeric(pos),
    TadStart = as.numeric(TadStart),
    TadEnd = as.numeric(TadEnd),
    DmrStart = as.numeric(DmrStart),
    DmrEnd = as.numeric(DmrEnd)
  )
```

# plot the cpgs linearly
```{r}
# Define x-axis tick positions for A and B regions definition
x_ticks <- c(95433599, 95479444, 95552720, 95533910, 95579759)

# Create the ggplot
p <- ggplot(cpgs_inCLDN10_AB, aes(x = pos, text = paste(Name, pos, sep = "\n"), color = GenType)) +  
  geom_linerange(aes(ymin = 0.95, ymax = 1.05)) +  # Small vertical lines with colors
  scale_x_continuous(breaks = x_ticks, labels = x_ticks) +  # Keep 5 x-axis labels
  scale_color_manual(values = c("CLDN10 Iso-A" = "blue", 
                                "CLDN10 Iso-B" = "red", 
                                "Targeted CpGs" = "green")) +  # Added green for Targeted CpGs
  labs(x = "Genomic Position CLDN10 (Chr13)", y = "", color = "CpG Type") +
  theme_minimal()

# Convert to an interactive plotly object
interactive_plot <- ggplotly(p, tooltip = "text")

# Save as HTML
saveWidget(interactive_plot, "./Plots/A_B_71CpGs_plot.html", selfcontained = TRUE, libdir = NULL)
```





# Add  die DMR und 2 TADs dazu
```{r}
library(ggplot2)
library(plotly)
library(htmlwidgets)
library(RColorBrewer)
library(dplyr)

# Define x-axis tick positions for A and B regions definition
x_ticks <- c(95433599, 95479444, 95552720, 95533910, 95579759)


# Ensure unique DMR_Nr values for coloring
unique_dmr_nrs <- unique(cpgs_inCLDN10_AB$DmrNr)
num_dmr_nrs <- length(unique_dmr_nrs)

# Define colors for DMRs
if (num_tad_nrs < 3) {
  color_palette <- RColorBrewer::brewer.pal(num_tad_nrs, "Set1")
} else {
  color_palette <- RColorBrewer::brewer.pal(num_tad_nrs, "Set3")
}



# Get the minimum CpG position (starting point of the plot)
min_pos <- min(cpgs_inCLDN10_AB$pos, na.rm = TRUE)
max_pos <- max(cpgs_inCLDN10_AB$pos, na.rm = TRUE)


# Create the plot
p2 <- ggplot(cpgs_inCLDN10_AB, aes(x = pos, text = paste(Name, pos, sep = "\n"), color = GenType)) +  

  # cpg lines
  geom_linerange(data = cpgs_inCLDN10_AB, 
                 aes(x = pos, ymin = 0, ymax = 1, color = GenType), 
                 size = 0.3) + 

  # Add vertical DMR rectangles, using color instead of fill
  geom_rect(data = cpgs_inCLDN10_AB[!is.na(cpgs_inCLDN10_AB$DmrNr), ],  
            aes(xmin = DmrStart, xmax = DmrEnd, ymin = 0, ymax = 1, color = DMR_label),  
            alpha = 0.4, inherit.aes = FALSE) +  

  # Add a vertical orange line at position 95500000
  geom_vline(xintercept = 95500000, color = "orange", linetype = "solid", size = 1) +  # Vertical line

  # Add text annotations before and after the vertical line
  annotate("text", x = 95500000 - 50000, y = 1.1, label = "Tad69", color = "orange", size = 4, hjust = 1) +  # Text before the line
  annotate("text", x = 95500000 + 50000, y = 1.1, label = "Tad70", color = "orange", size = 4, hjust = 0) +  # Text after the line
  annotate("text", x = 95500000, y = -0.06, label = "95500000", color = "orange", size = 3, hjust = 0.5) +  # Text on the x-axis    # X-axis settings
  
  scale_x_continuous(breaks = x_ticks, labels = x_ticks) +  

  # Color settings
  scale_color_manual(values = c("CLDN10 Iso-A" = "blue", 
                                "CLDN10 Iso-B" = "red", 
                                "Targeted CpGs" = "green")) +  

  # Labels and theme
  labs(x = "Genomic Position CLDN10 (Chr13)", y = "", 
       color = "CpG Type", fill = "DMRS") +  
  theme_minimal() 

# Convert to an interactive plotly object
interactive_p2 <- ggplotly(p2, tooltip = "text")



saveWidget(interactive_p2, "./Plots/A_B_71CpGs_plot_DMR+Tad_labels5.html", selfcontained = TRUE, libdir = NULL)
```
